com.unity.netcode.gameobjects/Tests/Runtime/DistributedAuthority/SessionVersionConnectionRequest.cs

```diff
@@ -0,0 +1,106 @@
+        protected override void OnNewClientCreated(NetworkManager networkManager)
+        {
+            m_ClientWasDisconnected = false;
+            m_ClientNetworkManager = networkManager;
+            m_ClientNetworkManager.OnClientDisconnectCallback += OnClientDisconnectCallback;
+            if (!m_UseValidSessionVersion)
+            {
+                networkManager.OnGetSessionConfig = GetInavlidSessionConfig;
+            }
+            base.OnNewClientCreated(networkManager);
+        }
+
+        /// <summary>
+        /// Tracks if the client was disconnected or not
+        /// </summary>
+        private void OnClientDisconnectCallback(ulong clientId)
+        {
+            m_ClientWasDisconnected = true;
+            m_ClientNetworkManager.OnClientDisconnectCallback -= OnClientDisconnectCallback;
+        }
+
+        /// <summary>
+        /// This handles disabling the internal integration test logic that waits for
+        /// clients to be connected. When we know the client will be disconnected,
+        /// we want to have the NetcodeIntegrationTest not wait for the client to
+        /// connect (otherwise it will timeout there and fail the test).
+        /// </summary>
+        /// <param name="networkManager"></param>
+        /// <returns>true = wait | false = don't wait</returns>
+        protected override bool ShouldWaitForNewClientToConnect(NetworkManager networkManager)
+        {
+            return m_UseValidSessionVersion;
+        }
+
+        /// <summary>
+        /// Validates that when the client's session config version is valid a client will be
+        /// allowed to connect and when it is not valid the client will be disconnected.
+        /// </summary>
+        /// <remarks>
+        /// This is just a mock of the service logic to validate everything on the NGO side is
+        /// working correctly.
+        /// </remarks>
+        /// <param name="useValidSessionVersion">true = use valid session version | false = use invalid session version</param>
+        [UnityTest]
+        public IEnumerator ValidateSessionVersion([Values] bool useValidSessionVersion)
+        {
+            // Test client being disconnected due to invalid session version
+            m_UseValidSessionVersion = useValidSessionVersion;
+            yield return CreateAndStartNewClient();
+            yield return s_DefaultWaitForTick;
+            if (!m_UseValidSessionVersion)
+            {
+                yield return WaitForConditionOrTimeOut(() => m_ClientWasDisconnected);
+                AssertOnTimeout("Client was not disconnected when it should have been!");
+                Assert.True(m_ClientNetworkManager.DisconnectReason == ConnectionRequestMessage.InvalidSessionVersionMessage, "Client did not receive the correct invalid session version message!");
+            }
+            else
+            {
+                Assert.False(m_ClientWasDisconnected, "Client was disconnected when it was expected to connect!");
```

Comment:
If the above check is using `WaitForConditionOrTimeOut()`, could this test give a false positive, where it checks if it wasn't disconnected before the connect is processed?

Reply:
By the time it reaches that point the clients will have already been verified connected. That wait condition is waiting for the client to notify it had been disconnected.